static/resources/harness.js

```diff
@@ -144,59 +177,62 @@
   //
   // If the test doesn't return true or false, or if it throws, `result` will
   // be null and a `message` property is set to an explanation.
-  function test(data) {
+  function test(data, callback) {
     var result = {name: data.name, info: {}};
 
-    for (var i = 0; i < data.tests.length; i++) {
+    function runTest(i, callback) {
+      function process(value) {
+        processValue(value, data, test, result);
+        if (result.result === true) {
+          callback(result);
+          return;
+        } else {
+          runTest(i + 1, callback);
+        }
+      }
+
+      if (i >= data.tests.length) {
+        callback(result);
+        return;
+      }
+
       var test = data.tests[i];
 
       try {
         var value = eval(test.code);
-        if (value && typeof value === 'object' && 'result' in value) {
-          result.result = value.result;
-          if (value.message) {
-            result.message = value.message;
-          }
-        } else if (typeof value === 'boolean') {
-          result.result = value;
+
+        if ('Promise' in self && value instanceof Promise) {
```

Comment:
Can we not instead detect a thenable here?

```js
if (typeof value === 'object' && typeof value.then === 'function') {
  value.then(process, process);
}
```

Writing that out, I also wonder what happens if a promise rejects with true or false? Seems like we should handle errors separately and make sure that always results in an inconclusive results (null) regardless of the rejection value.

Reply:
You're right!  I'll update the code to make the rejection create an error object (which'll create a null result with the error as the message).

Regarding the detection method, I'm not sure what additional benefit that would grant us (at least for the time being), but I suppose that we could do it like that, yeah!

Reply:
And I just realized the redundant use of `Promise.resolve()`, haha, whoops!